### PR TITLE
verbose.mk: fallback to standard file descriptors

### PR DESCRIPTION
**Pull request by mcprat**
ping @aparcar @guidosarducci 

In order to pass a status message at runtime,
which is usually listing subtargets
of a Makefile target or an error message,
from a sub-invocation of Make (submake)
to the top level invocation of Make,
the file descriptors 8 and 9 are opened to be used by the functions MESSAGE and ERROR_MESSAGE.

However, there are situations where those functions can be called while not in a submake, or for whatever reason, the file descriptors have not or cannot be opened, which results in a shell error:

/bin/bash: 8: Bad file descriptor

Commit aee3594ffcb72ae3e18c3719012d52519ee2d160
("verbose.mk: print ERROR messages in non-verbose") has exposed this issue to more cases, but it is not the root cause.

To solve this, use the exit code of the first printf attempt to the alternative file descriptors in order to tell whether the standard file descriptors need to be used instead. In order to get rid of the error printing, stderr is redirected to null after grouping the two printf alternatives into one command to combine outputs.

For ERROR_MESSAGE, the real message is redirected to stderr after redirecting the error from the attempted printing to null.

For MESSAGE, because the Make function "shell" absorbs stdout in order to input the value from stdout into the Makefile, the dummy variable "_NULL", previously used merely for causing a call to the MESSAGE function to trigger without writing target rules, now has a specific use case and a real value, so rename it to "_MESSAGE" as a placeholder for the real message when it must be printed to stdout.

When "_MESSAGE" has a value, use Make function "info" to finally bring it to the terminal.

This also fixes what is likely a typo, in that
while file descriptor 9 redirects to stderr for use in the function ERROR_MESSAGE for errors, that function has printf redirecting to file descriptor 8 instead.

### CODE REVIEW

**Comment by mpratt14** [comment]
a quick way to test

```
make package/symlinks
make package/symlinks >/dev/null
```

and then in variable `_MESSAGE` change the call to `ERROR_MESSAGE` and repeat

**Comment by guidosarducci** [comment]
A couple of comments. 

> However, there are situations where those functions can be called while not in a submake, or for whatever reason, the file descriptors have not or cannot be opened, which results in a shell error:

We really need to understand those reasons.

> To solve this, use the exit code of the first printf attempt to the alternative file descriptors in order to tell whether the standard file descriptors need to be used instead. In order to get rid of the error printing, stderr is redirected to null after grouping the two printf alternatives into one command to combine outputs.

This is basically trying different file descriptors until one works, and that seems more of a kludge than a proper fix for the underlying issue. Better perhaps to first understand where and why `make` invocations are doing unexpected redirections.

Honestly, since we only use `MESSAGE` and `ERROR_MESSAGE` in 4 places, and given the amount of time ~~wasted~~ spent by multiple people on this, I'm tempted to suggest just removing it. The small benefit (e.g. colour) seems outweighed by its complexity.

I'm hoping @aparcar has a cleaner solution in mind...

> This also fixes what is likely a typo, in that while file descriptor 9 redirects to stderr for use in the function ERROR_MESSAGE for errors, that function has printf redirecting to file descriptor 8 instead.

I agree that fd 8 and 9 were probably intended as analogs of stdout and stderr respectively, and fixed that in my attempts also. 

Some of the PR checks are failing too. Do you know what that's about?

**Comment by mpratt14** [comment]
any makefile target that is ran in a submake has the possibility of being called at the top level, in that case the user themselves would have to open up the file descriptors 8 and 9 themselves by typing it, unless maybe the command line itself can be overridden...

**Comment by mpratt14** [comment]
there might be an exit code issue, I had previously grouped the printf in parentheses but switched to braces to not spawn a subshell...

**Comment by mpratt14** [comment]
to be honest, lets give removing the file descriptors a try

I believe it was first done because of the Make shell function which absorbed stdout, only for the purpose of handling the MESSAGE function before we had ERROR_MESSAGE

**Comment by mpratt14** [comment]
original commit is still available to look at

https://github.com/openwrt/openwrt/commit/286b78019c10994a94433ca85c65534bb5b9d906

**Comment by mpratt14** [comment]
I cleaned up the commit message

the background of this problem is as old as the whole project... in 2006

https://github.com/openwrt/openwrt/commit/21ba6dba2643fcadcbba39523d5f4540a8d9aadd

I'm assuming it was inspired by subshell problems, like this

https://stackoverflow.com/questions/15153158/how-to-redirect-an-output-file-descriptor-of-a-subshell-to-an-input-file-descrip

**Comment by mpratt14** [comment]
getting rid of the file descriptor redirections gets rid of the messages altogether so I'm going back to what I had

**Comment by mpratt14** [comment]
> given the amount of time spent by multiple people on this, I'm tempted to suggest just removing it. The small benefit (e.g. colour) seems outweighed by its complexity.

I gave that alternative a decent try. my experiment with removing the file descriptor redirections made the messages just disappear for the currently working cases, for example running a simple `make clean` has no output...

you can still check out that commit here https://github.com/openwrt/openwrt/commit/aa6c60a1927b03f7a320dc9ea671009624987fd7

So the purpose is deeper than just being able to print in color

and as long as we learn something, no time is wasted 😄 

> seems more of a kludge than a proper fix for the underlying issue. Better perhaps to first understand where and why make invocations are doing unexpected redirections

This might be a cleaner solution than what it appears to be at first glance, and the redirections are not unexpected, rather there just happens to be an unexpected case where redirections are not needed. The best way to explain it probably is, these MESSAGE and ERROR_MESSAGE functions were originally written to be used _exclusively_ for child make processes in order to pass output to the parent make process, and after this change, they can be called from both without an issue.

anyway, I would say this is ready for merge now

**Comment by mpratt14** [comment]
CI won't tell us what's actually going on...

**Comment by mpratt14** [comment]
I found the problem

in scripts/json_overview_image_info.py

make is called, and probably the message is now printing instead of "bad file descriptor" to stderr, causing a larger than expected number of values to be input into python

I wonder if there's a better way to do this:

```
if output:
    default_packages, output["arch_packages"] = run(
        [
            "make",
            "--no-print-directory",
            "-C",
            "target/linux/",
            "val.DEFAULT_PACKAGES",
            "val.ARCH_PACKAGES",
        ],
        stdout=PIPE,
        stderr=PIPE,
        check=True,
        env=environ.copy().update({"TOPDIR": Path().cwd()}),
        universal_newlines=True,
    ).stdout.splitlines()
```

**Comment by mpratt14** [comment]
the regression is now being handled

I looked for other scripts which run make that might need adjustments, but no others seem to be affected, or they print the output to terminal so it doesn't matter

as far as I can tell, only the val.% targets are being used when using the output of make as an input to something else

how does it look now? @hauke @aparcar

**Comment by mpratt14** [comment]
just changed parentheses back to braces so there's no subshell

File: include/toplevel.mk [hunk 1]
```diff
@@ -215,6 +215,9 @@ check: .config FORCE
 val.%: FORCE
 	@+$(NO_TRACE_MAKE) -r -s $@ QUIET= V=s
 
+var.%: FORCE
+	@+$(NO_TRACE_MAKE) -r -s $@ QUIET= V=s
+
```
Comment (guidosarducci) [inline include/toplevel.mk:220]: Couldn't these targets simply be combined? Makes it obvious their handling is the same... 
```
val.% var.%: FORCE
	@+$(NO_TRACE_MAKE) -r -s $@ QUIET= V=s
```

Comment (mpratt14) [reply include/toplevel.mk]: yeah good idea, I reduced it to a 1 line diff

**Comment by mpratt14** [comment]
any more ideas? does this look ready?

File: include/toplevel.mk [hunk 2]
```diff
@@ -213,6 +213,7 @@ check: .config FORCE
 	@+$(NO_TRACE_MAKE) -r -s $@ QUIET= V=s
 
 val.%: FORCE
+var.%: FORCE
```
Comment (guidosarducci) [inline include/toplevel.mk:216]: Why not merge these as noted before? Doesn't this break the `val.%` target?

Comment (mpratt14) [reply include/toplevel.mk:216]: yeah you're right, I don't know why I thought I had done something like that before...

**Comment by mpratt14** [comment]
hey @Ansuel 

looking at the CI error, it's using the newest libubox checkout, but this branch doesn't have that commit at all... it's in master only 

I was going to rebase but I wanted you to see first

**Comment by Ansuel** [comment]
@mpratt14 unrelated

**Comment by mpratt14** [comment]
Yeah, it's unrelated, but I thought CI checked out the branch instead of rebase on master....

**Comment by mpratt14** [comment]
rebased, ready for review...

**Comment by guidosarducci** [comment]
@mpratt14 @Ansuel We should revert the change to the `check` target as this is unrelated to the `val.%` / `var.%` change, confuses history, and mistakenly drops a `.config` dependency. It also generates a ton of make warnings during package updates, and thus appears untested:
```
/home/kodidev/openwrt-project/include/toplevel.mk:213: *** mixed implicit and normal rules: deprecated syntax
/home/kodidev/openwrt-project/include/toplevel.mk:213: *** mixed implicit and normal rules: deprecated syntax
/home/kodidev/openwrt-project/include/toplevel.mk:213: *** mixed implicit and normal rules: deprecated syntax
/home/kodidev/openwrt-project/include/toplevel.mk:213: *** mixed implicit and normal rules: deprecated syntax
...
```

The patch below works for me:
```
diff --git a/include/toplevel.mk b/include/toplevel.mk
index 96f32139d9..f2dfde60cf 100644
--- a/include/toplevel.mk
+++ b/include/toplevel.mk
@@ -210,7 +210,9 @@ prereq:: prepare-tmpinfo .config
        @+$(NO_TRACE_MAKE) -r -s $@

 check: .config FORCE
-check val.% var.%: FORCE
+       @+$(NO_TRACE_MAKE) -r -s $@ QUIET= V=s
+
+val.% var.%: FORCE
        @+$(NO_TRACE_MAKE) -r -s $@ QUIET= V=s

 WARN_PARALLEL_ERROR = $(if $(BUILD_LOG),,$(and $(filter -j,$(MAKEFLAGS)),$(findstring s,$(OPENWRT_VERBOSE))))

```

**Comment by mpratt14** [comment]
Yeah sorry, comments on the commit also described this.

https://github.com/openwrt/openwrt/commit/e4a43cda0efba562d0ed1973b584c21309040ccc

 I thought CI or someone else would catch any issues beforehand...

We're just waiting for someone to push the patch, or does it help to make a PR?

**Comment by mpratt14** [comment]
By the way, the dependency for `.config` is still there, splitting dependencies between different lines is additive.

The only problem is mixing normal target names with patterns...

**Comment by guidosarducci** [comment]
But it's also confusing for maintenance and needlessly conflates histories of `check` and `val/var` targets...

Feel free to respin or reword that PR; I'd just like to see the problem resolved and so posted a patch.

**Comment by mpratt14** [comment]
for future reference the following works...
Although I agree that it's much less readable, so I wouldn't bother trying to do this here, but it's nice to know anyway...

```
 check: .config FORCE
$(filter check val.% var.%,$(MAKECMDGOALS)): FORCE
        @+$(NO_TRACE_MAKE) -r -s $@ QUIET= V=s
```

**Pull request merged**